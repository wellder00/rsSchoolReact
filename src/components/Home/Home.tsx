import styles from './Home.module.scss';

import { ErrorBoundary } from '../../components/ErrorBoundary';
import { Header } from '../../components/Header';
import { Main } from '../../components/Main';

// import { useEffect } from 'react';
import { useAppDispatch } from '@/lib/redux/hooks/reduxHooks';
import { changeItemsAmount } from '@/lib/redux/slices/itemsPerPageSlice';

const Home = () => {
  const dispatch = useAppDispatch();
  // const currentPage = useAppSelector((state) => state.itemsAmount.currentPage);

  // const [, setParams] = useSearchParams();

  // useEffect(() => {
  //   setParams({
  //     page: currentPage,
  //   });
  // }, [currentPage]);

  const onSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) =>
    dispatch(changeItemsAmount(event.target.value));

  return (
    <div className={styles.wrapper}>
      <ErrorBoundary>
        <Header onSelectChange={onSelectChange} />
      </ErrorBoundary>
      <Main />
    </div>
  );
};

export default Home;
