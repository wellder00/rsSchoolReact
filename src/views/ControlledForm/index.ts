export { default as ControlledForm } from './ControlledForm';
