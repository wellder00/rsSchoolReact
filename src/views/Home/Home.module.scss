.wrapper {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  min-height: 100vh;
}