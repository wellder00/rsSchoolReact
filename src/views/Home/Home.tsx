import { useEffect, useState } from 'react';
import { useNavigate, useSearchParams } from 'react-router-dom';

import styles from './Home.module.scss';

import pokemonDataContext from '../../state/ContextPokemonData';
import { useAppDispatch, useAppSelector } from '../../Hooks/reduxHooks';

import { Header } from '@components/Header';
import { Main } from '@components/Main';
import { ErrorBoundary } from '@components/ErrorBoundary';

import { getPokemon } from '../../api/api';
import { MyContextType, Pages } from 'types/interfaces';
import { changeItemsAmount } from '../../store/itemsPerPageSlice';
import { useGetPokemonsQuery } from '../../store/redux/pokemonApi';

const Home = () => {
  const dispatch = useAppDispatch();
  const itemsAmount = useAppSelector((state) => state.itemsAmount.items);
  const { data = [] } = useGetPokemonsQuery({ limit: itemsAmount, offset: '0' });

  console.log(data);

  const [pokemonData, setPokemonData] = useState<MyContextType>(data);
  const [searchParams, setSearchParams] = useSearchParams();
  const initialValueLimit = searchParams.get('limit') || itemsAmount;
  const initialValueOffset = searchParams.get('offset') || 0;
  const initialValuePage = searchParams.get('page') || 1;
  const [pages, setPages] = useState<Pages>({
    offset: +initialValueOffset,
    currentPage: +initialValuePage,
    lastPage: 2,
  });
  const navigate = useNavigate();
  const url = `/?limit=${initialValueLimit}&offset=0&page=1`;

  async function fetchData(pokemon: string, offset: number, limit: number) {
    try {
      const data = await getPokemon(pokemon, offset, limit);
      setPokemonData(data);
    } catch (error) {
      console.error(error);
    }
  }
  useEffect(() => {
    setSearchParams({
      limit: '' + itemsAmount,
      offset: '' + pages.offset,
      page: '' + pages.currentPage,
    });
    fetchData('', pages.offset, +itemsAmount);
  }, [pages]);

  useEffect(() => {
    if (searchParams.get('limit') !== itemsAmount) {
      setSearchParams({ limit: '' + itemsAmount, offset: '0', page: '1' });
      setPages({ ...pages, currentPage: 1, offset: 0 });
    } else {
      setSearchParams({
        limit: '' + itemsAmount,
        offset: '' + pages.offset,
        page: '' + pages.currentPage,
      });
      fetchData('', pages.offset, +itemsAmount);
    }
  }, [itemsAmount]);

  // useEffect(() => {
  //   dispatch(changeItemsAmount(initialValueLimit));
  //   async function fetchData(pokemon: string) {
  //     const data = await getPokemon(pokemon);
  //     setPokemonData(data);
  //   }

  //   const storedData = localStorage.getItem('pokemon');
  //   if (storedData) {
  //     const parsedData = JSON.parse(storedData);
  //     fetchData(parsedData);
  //   } else {
  //     fetchData('');
  //   }
  // }, []);

  const findCharacter = async (character: string) => {
    try {
      const data = await getPokemon(character);
      setPokemonData(data);
      navigate(url);
    } catch (error) {
      console.error(error);
    }
  };

  const onSelectChange = (event: React.ChangeEvent<HTMLSelectElement>) =>
    dispatch(changeItemsAmount(event.target.value));

  const onChangePage = (isNext: boolean) => {
    const limit = searchParams.get('limit');
    const offset = searchParams.get('offset');
    if (pokemonData?.count && limit && offset) {
      const lastPage = Math.ceil(pokemonData.count / +limit);
      const offsetChange = isNext ? +limit : -limit;
      const newOffset = parseInt(offset) + offsetChange;
      const currentPage = Math.floor(newOffset / +limit) + 1;

      setPages({ currentPage, lastPage, offset: newOffset });
    }
  };

  return (
    <pokemonDataContext.Provider value={pokemonData}>
      <div className={styles.wrapper}>
        <ErrorBoundary>
          <Header findCharacter={findCharacter} onSelectChange={onSelectChange} />
        </ErrorBoundary>
        <Main onChangePage={onChangePage} pages={pages} />
      </div>
    </pokemonDataContext.Provider>
  );
};

export default Home;
