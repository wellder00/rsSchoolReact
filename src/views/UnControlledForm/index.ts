export { default as UnControlledForm } from './UnControlledForm';
