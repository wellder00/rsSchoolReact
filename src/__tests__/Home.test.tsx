import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Home } from '../views/Home';

describe('Tests for the Card component', () => {
  it('Testing home page', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText(/Loading.../i)).toBeInTheDocument();
  });
});
