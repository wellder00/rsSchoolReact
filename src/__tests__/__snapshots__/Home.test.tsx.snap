// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for the Card component > Check that an appropriate message is displayed if no cards are present 1`] = `
<DocumentFragment>
  <div
    class="_wrapper_81be75"
  >
    <div
      class="_wrapper_f4f7a7"
    >
      <img
        alt="title"
        class="_titleImg_f4f7a7"
        src="/src/assets/images/title.png"
      />
      <div
        class="_searchWrap_f4f7a7"
      >
        <input
          class="_search_823600"
          placeholder="search"
          type="text"
          value=""
        />
        <button
          class="_search_f8f296 _button_f8f296"
        >
          SEARCH
        </button>
        <button
          class="_error_f8f296 _button_f8f296"
        >
          TRY MAKE ERROR
        </button>
      </div>
      <img
        alt="logo"
        class="_logo_f4f7a7"
        src="/src/assets/images/logo.png"
      />
    </div>
    <div
      class="_wrapper_7a8146"
    >
      <img
        alt="teleportTop"
        class="_teleportTop_7a8146"
        src="/src/assets/images/teleportTop.png"
      />
      <div
        class="_infoAndCardWrap_7a8146"
      >
        <div
          class="_wrapCard_7a8146"
        >
          <div
            class="_loading_d51639"
          >
            <div
              class="_loadingIcon_d51639"
              data-testid="loading-icon"
            >
              Loading...
            </div>
            <div
              class="_loader_d51639"
            >
              <div
                class="_loaderCircle_d51639"
              />
            </div>
          </div>
        </div>
      </div>
      <img
        alt="teleportBottom"
        class="_teleportBottom_7a8146"
        src="/src/assets/images/teleportBottom.png"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tests for the Card component > Testing home page 1`] = `
<DocumentFragment>
  <div
    class="_wrapper_81be75"
  >
    <div
      class="_wrapper_f4f7a7"
    >
      <img
        alt="title"
        class="_titleImg_f4f7a7"
        src="/src/assets/images/title.png"
      />
      <div
        class="_searchWrap_f4f7a7"
      >
        <input
          class="_search_823600"
          placeholder="search"
          type="text"
          value=""
        />
        <button
          class="_search_f8f296 _button_f8f296"
        >
          SEARCH
        </button>
        <button
          class="_error_f8f296 _button_f8f296"
        >
          TRY MAKE ERROR
        </button>
      </div>
      <img
        alt="logo"
        class="_logo_f4f7a7"
        src="/src/assets/images/logo.png"
      />
    </div>
    <div
      class="_wrapper_7a8146"
    >
      <img
        alt="teleportTop"
        class="_teleportTop_7a8146"
        src="/src/assets/images/teleportTop.png"
      />
      <div
        class="_infoAndCardWrap_7a8146"
      >
        <div
          class="_wrapCard_7a8146"
        >
          <div
            class="_loading_d51639"
          >
            <div
              class="_loadingIcon_d51639"
              data-testid="loading-icon"
            >
              Loading...
            </div>
            <div
              class="_loader_d51639"
            >
              <div
                class="_loaderCircle_d51639"
              />
            </div>
          </div>
        </div>
      </div>
      <img
        alt="teleportBottom"
        class="_teleportBottom_7a8146"
        src="/src/assets/images/teleportBottom.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
